File: python/274_HIndex.py
from unittest import TestCase
# https://leetcode.com/problems/h-index/


class HIndex(TestCase):
    def hIndex(self, citations):
        """
        :type citations: List[int]
        :rtype: int
        """
        n = len(citations)
        cnt = [0 for _ in range(n)]
        for c in citations:
            if c < n: cnt[c] += 1
        at_least = n
        for i, c in enumerate(cnt):
            if at_least < i: return i-1
            at_least -= c
        return n - (at_least < n)

    def test1(self):
        self.assertEqual(3, self.hIndex([3,0,6,1,5]))

File: kotlin/src/main/kotlin/274.HIndex.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/h-index/


class HIndex{
  fun hIndex(citations: IntArray): Int {
    val n = citations.size
    val cnt = IntArray(n+1)
    for(c in citations)
      cnt[minOf(c, n)] ++
    var sum = 0
    for(i in n downTo 1){
      sum += cnt[i]
      if(sum >= i)
        return i
    }
    return 0
  }

  @Test
  fun test1(){
    assertEquals(3, hIndex(intArrayOf(3,0,6,1,5)))
  }

  @Test
  fun test2(){
    assertEquals(0, hIndex(intArrayOf()))
  }
}